File: core/src/com/maattss/intro/exercise/sprites/Ball.java
package com.maattss.intro.exercise.sprites;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.maattss.intro.exercise.states.Task4State;

public class Ball {
    private Texture ball;
    private Vector2 position;
    private boolean upHeading;
    private boolean rightHeading;
    private int speed;

    private Rectangle bounds;


    public Ball(int x, int y) {
        position = new Vector2(x,y);
        ball = new Texture("ball.png");
        upHeading = true;
        rightHeading = true;
        speed = 600;
        bounds = new Rectangle(position.x,position.y,20,20);
    }

    public void update(float dt, Task4State ps, PaddleLeft pl, PaddleRight pr) {
        bounds.setPosition(position.x,position.y);

        if(bounds.overlaps(pl.getBounds()) || bounds.overlaps(pr.getBounds())){
            this.rightHeading=!this.rightHeading;
        }

        this.moveY(speed*dt);
        this.moveX(speed*dt, ps);
    }

    public Rectangle getBounds() {
        return bounds;
    }

    public void moveX(float speed, Task4State ps) {
        if(rightHeading) {
            position.x += speed;
        }
        else {
            position.x -= speed;
        }


        if(position.x > 600) { //         if(rightHeading && (position.x> (IntroExercise.WIDTH - ball.getWidth()))) {
            position.x = 290;
            position.y = 290;
            rightHeading=false;
            upHeading=!upHeading;
            ps.incLeftScore();
        }
        if(position.x < 0) { //         if(rightHeading && (position.x> (IntroExercise.WIDTH - ball.getWidth()))) {
            position.x = 290;
            position.y = 290;
            rightHeading=true;
            upHeading=!upHeading;
            ps.incRightScore();
        }

    }

    public void moveY(float speed) {
        if(upHeading) {
            position.y += speed;
        }
        else {
            position.y -= speed;
        }

        if(upHeading && (position.y > 580)) { //         if(upHeading && (position.y > (IntroExercise.HEIGHT - ball.getHeight()))) {
            position.y -= speed;
            upHeading=false;
        }

        if(!upHeading && (position.y < 0)) { //         if(upHeading && (position.y > (IntroExercise.HEIGHT - ball.getHeight()))) {
            position.y += speed;
            upHeading=true;
        }
    }

    public Vector2 getPosition() {
        return position;
    }

    public Texture getTexture() {
        return ball;
    }
}

File: core/src/com/maattss/intro/exercise/sprites/Ball.kt
package com.maattss.intro.exercise.sprites

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.maattss.intro.exercise.IntroExercise
import com.maattss.intro.exercise.states.Task4State

class Ball {
    val texture: Texture = Texture("pong/ball.png")
    val position: Vector2 = Vector2((IntroExercise.WIDTH / 2).toFloat(),
            (IntroExercise.HEIGHT / 2).toFloat())
    private var up: Boolean = true
    private var right: Boolean = true
    private val speed: Int = 400
    val bounds: Rectangle = Rectangle(position.x, position.y, 20f, 20f)

    fun update(dt: Float, state: Task4State, pl: LeftPaddle, pr: RightPaddle) {
        bounds.setPosition(position.x, position.y)
        if (bounds.overlaps(pl.bounds) || bounds.overlaps(pr.bounds)) {
            right = !right
        }
        moveY(speed * dt)
        moveX(speed * dt, state)
    }

    fun moveX(speed: Float, state: Task4State) {
        if (right) {
            position.x += speed
        } else {
            position.x -= speed
        }
        if (position.x > IntroExercise.WIDTH) {
            position.x = IntroExercise.WIDTH / 2.toFloat()
            position.y = IntroExercise.HEIGHT / 2.toFloat()
            right = false
            up = !up
            state.incLeftScore()
        }
        if (position.x < 0) {
            position.x = IntroExercise.WIDTH / 2.toFloat()
            position.y = IntroExercise.HEIGHT / 2.toFloat()
            right = true
            up = !up
            state.incRightScore()
        }
    }

    fun moveY(speed: Float) {
        if (up) {
            position.y += speed
        } else {
            position.y -= speed
        }
        if (up && position.y > IntroExercise.HEIGHT) up = false
        if (!up && position.y < 0) up = true
    }
}